.github/workflows/check-url.sh

```diff
@@ -6,23 +6,32 @@ fi
 
 filename=$(cut -d ";" -f 1 <<< "$1")
 URL=$(cut -d ";" -f 2 <<< "$1")
+# --head: Sends a HEAD request. We use this to be good netizens, since we only need the header to check the response code.
+# --silent: Hides the curl progress bar, which is unnecessary noise when testing >600 urls.
+# --insecure: Skip SSL verification. Since we're only checking the headers, this should be safe to do. (https://curl.se/docs/sslcerts.html)
+CURL_ARGS=(--head --silent --insecure)
+# Explicitly write out the HTTP code to stdout, while redirecting the response output to /dev/null
+HTTP_CODE_ONLY=(--write-out '%{http_code}' --output /dev/null)
+# Override default behavior (exponential backoff + 10m limit) since we don't need that many retries
+# We still keep a 5m limit, though, because --retry respects the Retry-After field, which may be greater than 30s.
+RETRY_ARGS=(--retry 2 --retry-delay 30 --retry-max-time 300)
 
 # Make sure to check both URL *and* redirections (--location) for excluded domains
-full_info=$(curl -I --silent --insecure --location -- "$URL")
-LOCATION=$(curl -I --silent --insecure -- "$URL" | perl -n -e '/^[Ll]ocation: (.*)$/ && print "$1\n"')
+full_info=$(curl "${CURL_ARGS[@]}" --location -- "$URL")
+LOCATION=$(curl "${CURL_ARGS[@]}" -- "$URL" | perl -n -e '/^[Ll]ocation: (.*)$/ && print "$1\n"')
 if [[ "$full_info + $URL" =~ 'drive.google.com'|'pipeline-hemis'|'sciencedirect.com'|'wiley.com'|'sagepub.com'|'ncbi.nlm.nih.gov'|'oxfordjournals.org'|'docker.com'|'ieeexplore.ieee.org'|'liebertpub.com'|'tandfonline.com'|'pnas.org'|'neurology.org'|'academic.oup.com'|'journals.lww.com'|'science.org'|'pubs.rsna.org'|'direct.mit.edu'|'archive.ph'|'mirror.centos.org'|'vault.centos.org'|'%s' ]]; then
```

Comment:
I wonder how many of the above domains only failed because we were trying to fetch the full text.

We may want to try a one-off test while removing all of these exceptions to see if they truly are proper errors, aha.

Reply:
Odd... I'm still seeing error code 429 for some domains, e.g. https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/11368764606/job/31624613634?pr=4649#step:7:520

That said, trying locally, I get an error code 429 even on my very first request, and the retry behavior is working correctly, so it looks like this is just a quirk of this specific domain:

```console
dot|joshua@mosaic:~$ curl --write-out '%{http_code}' --insecure --retry 3 --retry-all-errors --retry-delay 30 -I -- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7019007/ 
HTTP/2 429 
strict-transport-security: max-age=31536000; includeSubDomains; preload
referrer-policy: origin-when-cross-origin
content-security-policy: upgrade-insecure-requests
date: Wed, 16 Oct 2024 15:46:10 GMT
server: nginx
refresh: 67
retry-after: 67
content-type: text/html
x-frame-options: DENY
x-ua-compatible: IE=Edge
x-xss-protection: 1; mode=block

Warning: Problem : HTTP error. Will retry in 67 seconds. 3 retries left.
HTTP/2 429 
strict-transport-security: max-age=31536000; includeSubDomains; preload
referrer-policy: origin-when-cross-origin
content-security-policy: upgrade-insecure-requests
date: Wed, 16 Oct 2024 15:47:18 GMT
server: nginx
refresh: 67
retry-after: 67
content-type: text/html
x-frame-options: DENY
x-ua-compatible: IE=Edge
x-xss-protection: 1; mode=block

Warning: Problem : HTTP error. Will retry in 67 seconds. 2 retries left.
```

retry-after is respected!! :tada:

Reply:
Sorting through the errored domains, I'm seeing:

<details>

```
⛔ Error status code: 400 for URL: https://github.com/spinalcordtoolbox/sct_tutorial_data/releases/download/r20241007/%s
⛔ Error status code: 403 for URL: http://journals.sagepub.com/doi/10.1177/20552173221132170
⛔ Error status code: 403 for URL: http://pubs.rsna.org/doi/abs/10.1148/radiol.2433060583
⛔ Error status code: 403 for URL: https://academic.oup.com/braincomms/article/doi/10.1093/braincomms/fcac246/6746720
⛔ Error status code: 403 for URL: https://academic.oup.com/braincomms/article/doi/10.1093/braincomms/fcad107/7103314
⛔ Error status code: 403 for URL: https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/10.1002/nbm.4964
⛔ Error status code: 403 for URL: https://analyticalsciencejournals.onlinelibrary.wiley.com/doi/10.1002/nbm.5115
⛔ Error status code: 403 for URL: https://brain.oxfordjournals.org/content/139/6/1735
⛔ Error status code: 403 for URL: https://direct.mit.edu/imag/article/doi/10.1162/imag_a_00075/119044/A-database-of-the-healthy-human-spinal-cord
⛔ Error status code: 403 for URL: https://direct.mit.edu/imag/article/doi/10.1162/imag_a_00075/119044/A-database-of-the-healthy-human-spinal-cord
⛔ Error status code: 403 for URL: https://direct.mit.edu/imag/article/doi/10.1162/imag_a_00218/122601/Automatic-segmentation-of-the-spinal-cord-nerve
⛔ Error status code: 403 for URL: https://direct.mit.edu/imag/article/doi/10.1162/imag_a_00225/123269/Optimization-of-cervical-cord-synthetic-T1
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/10.1177/0271678X20982396
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/10.1177/13524585221115304
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/10.1177/13524585241229969
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/10.1177/20552173231211396
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/full/10.1177/10738584221101827
⛔ Error status code: 403 for URL: https://journals.sagepub.com/doi/pdf/10.1177/1756286419840593
⛔ Error status code: 403 for URL: https://movementdisorders.onlinelibrary.wiley.com/doi/10.1002/mds.29169
⛔ Error status code: 403 for URL: https://movementdisorders.onlinelibrary.wiley.com/doi/10.1002/mds.29261
⛔ Error status code: 403 for URL: https://movementdisorders.onlinelibrary.wiley.com/doi/10.1002/mds.29354
⛔ Error status code: 403 for URL: https://movementdisorders.onlinelibrary.wiley.com/doi/10.1002/mds.29934
⛔ Error status code: 403 for URL: https://online.liebertpub.com/doi/abs/10.1089/neu.2017.4980
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/ana.24932
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/hbm.26018
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/hbm.26523
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/hbm.26600
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/ima.22859
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.26852/full
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.26909/full
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.26945/full
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.27105/full
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.29461
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.29708
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/mrm.30137
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1002/nbm.3801/full
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1111/ene.15027
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1111/jon.13076
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1111/jon.13240
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1111/jvim.16014
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/10.1155/2014/719520
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/abs/10.1002/ana.25571
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/abs/10.1002/jmri.27253
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/epdf/10.1002/jmri.26269
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/epdf/10.1002/mrm.28751
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/full/10.1002/nbm.4142
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/full/10.1111/os.13326
⛔ Error status code: 403 for URL: https://onlinelibrary.wiley.com/doi/full/10.1111/vru.13126
⛔ Error status code: 403 for URL: https://twitter.com/spinaltoolbox
⛔ Error status code: 403 for URL: https://www.docker.com/what-container/
⛔ Error status code: 403 for URL: https://www.docker.com/what-container/
⛔ Error status code: 403 for URL: https://www.docker.com/what-container/
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neu.2022.0034
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neu.2022.0283
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neu.2023.0305
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neu.2023.0591
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neu.2024.0267
⛔ Error status code: 403 for URL: https://www.liebertpub.com/doi/10.1089/neur.2020.0035
⛔ Error status code: 403 for URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7061306/
⛔ Error status code: 403 for URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8039452/
⛔ Error status code: 403 for URL: https://www.neurology.org/doi/10.1212/WNL.0000000000008318
⛔ Error status code: 403 for URL: https://www.neurology.org/doi/10.1212/WNL.0000000000012643
⛔ Error status code: 403 for URL: https://www.neurology.org/doi/10.1212/WNL.0000000000201465
⛔ Error status code: 403 for URL: https://www.neurology.org/doi/abs/10.1212/WNL.0000000000012643
⛔ Error status code: 403 for URL: https://www.pnas.org/content/111/50/18067.abstract
⛔ Error status code: 403 for URL: https://www.science.org/doi/10.1126/sciadv.adh9853
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/journal/seminars-in-ultrasound-ct-and-mri/articles-in-press
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/abs/pii/S0967586820316192
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S0003999317312844
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S0035378720306159
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S0168010222001894
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S0306452217307637
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S088721711630049X
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811916305560
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811917302185#f0005
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811917305220
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811917308686
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811917308686
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811917308686
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811918319578
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811918319578
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811919304288
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811920303918
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1053811920309241?via%3Dihub
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S136184151730186X
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S1878875019308459
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S221315821730150X
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S2213158218303668
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S2213158219303122
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S2213158221000103
⛔ Error status code: 403 for URL: https://www.sciencedirect.com/science/article/pii/S221315822100293X
⛔ Error status code: 403 for URL: https://www.tandfonline.com/doi/full/10.1080/10255842.2015.1069625
⛔ Error status code: 403 for URL: https://www.tandfonline.com/doi/full/10.1080/10790268.2018.1527079
⛔ Error status code: 404 for URL: https://github.com/ivadomed/pipeline-hemis/issues/3
⛔ Error status code: 418 for URL: https://ieeexplore.ieee.org/document/9522652
⛔ Error status code: 429 for URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6684621/
⛔ Error status code: 429 for URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7019007/
⛔ Error status code: 500 for URL: https://www.spiedigitallibrary.org/conference-proceedings-of-spie/12032/1203213/Mapping-pediatric-spinal-cord-development-with-age/10.1117/12.2612210.short?SSO=1
```

</details>

So, the list of erroring domains is still largely centered on publishers. But, I think there's a chance that some of the originally-excluded domains are actually valid. So, I'll go ahead and update the list. :)

Reply:
It looks like two of those domains are newly failing (see https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4649/commits/0b904bb2fb0d7095d1f4cf8f4b447259ed5013e5 and https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/11369554032/job/31627268516).

With the newly-added exceptions, we should be good to go. :tada: